gms/gossiper.cc

```diff
@@ -754,9 +754,31 @@ future<> gossiper::update_live_endpoints_version() {
     auto version = _live_endpoints_version + 1;
     return container().invoke_on_all([version] (gms::gossiper& g) {
         g._live_endpoints_version = version;
+        g._live_endpoints_version_condvar.broadcast();
     });
 }
 
+future<std::set<inet_address>> gossiper::get_live_members_synchronized(std::chrono::milliseconds timeout) {
+    auto live_members = gossiper::get_live_members();
+    auto version = _live_endpoints_version;
+    size_t nr_version = co_await container().map_reduce0([version, timeout] (gossiper& local_gossiper) -> size_t {
+        auto f = local_gossiper._live_endpoints_version_condvar.wait(timeout, [version, &local_gossiper] {
+             return local_gossiper._live_endpoints_version >= version;
+        });
+        f.get();
+        if (f.failed()) {
+            f.ignore_ready_future();
+            return 0;
+        }
+        return 1;
+    }, 0, std::plus<size_t>());
+    logger.debug("Nodes with gossiper version >= {}: {} / {}", version, nr_version, smp::count);
+    if (nr_version < smp::count) {
+        throw std::runtime_error(format("Failed wait for gossip version on all shards in 20ms"));
+    }
+    co_return live_members;
```

Comment:
I think you want to return when get_live_members (_live_endpoints) returns the same set of nodes on all shards. But the same _live_endpoints_version on all shards does not guarantees _live_endpoints is the same on all shards.

Reply:
In gossiper::run, we can add a condition_variable and signal it when _live_endpoints is synced from shard zero to other shards make gossiper::get_live_members_synchronized wait for it.

Reply:
Or,  check / sleep/ retry in a loop until times out.

Or, simpler, just return the live members from all shards immediately, it is up to the user to retry
{
shard : 0, live_members : [n1, n2, n3]
shard : 1, livemembers : [n1, n2 ]
}

Reply:
We don't want this. There is no point in waiting for them all to be equal.

Reply:
> I think you want to return when get_live_members (_live_endpoints) returns the same set of nodes on all shards. But the same _live_endpoints_version on all shards does not guarantees _live_endpoints is the same on all shards.

No, we want to make sure that before we return the status from shard 0 all nodes saw this status. Liveness information can be outdated immediately after it is acquired anyway, so there is no point to wait for all sets to be the same. It may not be the case  nanosecond after we returned the result anyway.

Reply:
OK. But as I mentioned earlier the current code does not make sure this.  Other shards see the same _live_endpoints_version does not mean other shards see the status of _live_endpoints from shard zero. 

One implementation could be:

In get live member function:

if (_shadow_live_endpoints == _live_endpoints) {
     return  _live_endpoints immediately // this means all shards see the same live endpoints
} else {
     wait for condition variable to check _shadow_live_endpoints == _live_endpoints
}

In gossiper::run where _live_endpoints is replicated to non zero shards:
signal condition variable after _live_endpoints of other shards are updated.

Reply:
It is not like _live_endpoints on N shards would have N different values. It is like we only have two set of values
shard 0  value  / non shard zero value, since all non zero shards will be updated in a batch. 

I do not see why waiting for all shards have the same _live_endpoints would be a problem.

Reply:
> OK. But as I mentioned earlier the current code does not make sure this. Other shards see the same _live_endpoints_version does not mean other shards see the status of _live_endpoints from shard zero.

Can you elaborate on that. Are you saying that _live_endpoints_version on a shard does not change each time _live_endpoints on the shard changes? If they change together like we assumed the code above does exactly what we want: checks that all shards saw a certain transition.

Reply:
The _live_endpoints_version was introduced for the direct gossip failure detector to re-calculate nodes each shard should be responsible for. 
 
See here:

```
future<> gossiper::update_live_endpoints_version() {
    auto version = _live_endpoints_version + 1;
    return container().invoke_on_all([version] (gms::gossiper& g) {
        g._live_endpoints_version = version;
    });
}
```

The _live_endpoints_version is updated but the _live_endpoints is not updated here. 

The actually place where _live_endpoints is updated is in the gossiper::run() loop.

Reply:
Then we need to introduce something that does what we want: changes together with _live_endpoints update. @kostja why did you think we can use `_live_endpoints_version` for that?

Reply:
I was also convinced that `_live_endpoints` get replicated together with `_live_endpoints_version` :(

Reply:
Indeed, they are only replicated in `run()`.
But `run()` is only happening once every second :(
Even if we implemented waiting based on this, it would slow down our test.

Why not replicate every time `_live_endpoints` changes on shard 0? They change infrequently anyway.